src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        // We attempt to make error recovery a little nicer here by only treating a single
+        // identifier token as a calling convention if the token afterwords is actually a
+        // possible start to the function pointer parameter list. The intuition is that, for
+        // some partial trees like this:
+        //
+        //  (delegate*MyProperty.MyMethod();
+        //
+        // The user is actually in the middle of adding a cast, and if we interpreted MyProperty
+        // as a calling convention, we'd get a much worse set of errors than if we treated the
+        // function pointer as having no calling convention.
+        // Also for nicer error recovery, we only treat predefined types as if they could be the
+        // calling convention. For any other keyword, they're almost certainly part of some other
+        // construct, and would produce better errors if parsed separately. The user could have this:
+        //
+        // delegate* // Declaration in progress
+        // while (true) {}
+        //
+        // The parse tree will look much better if the while(true) is considered a separate structure,
+        // even though it does look like it could be the start of an invalid function pointer definition.
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```

Comment:
>PeekToken(1) [](start = 59, length = 12)

Is it always safe to peek one token ahead?

Reply:
Yes, I'll add a test. If we're at the end of file, we'll get back more end of file tokesn.

---
In reply to: [349359044](https://github.com/dotnet/roslyn/pull/39879#discussion_r349359044) [](ancestors = 349359044)